sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -672,4 +689,300 @@ private WindowDoFnAdapter(DoFnReflector reflector, DoFnWithContext<InputT, Outpu
       super(reflector, fn);
     }
   }
+
+  /**
+   * A byte-buddy {@link Implementation} that delegates a call that receives
+   * {@link ExtraContextInfo} to the given {@link DoFnWithContext} method.
+   */
+  private static final class InvokerDelegation implements Implementation {
+    @Nullable
+    private final Method target;
+    private final boolean isStartBundle;
+    private final ExtraContextInfo[] args;
+    private final Assigner assigner = Assigner.DEFAULT;
+    private FieldDescription field;
+
+    /**
+     * Create the {@link InvokerDelegation} for the specified method.
+     *
+     * @param target the method to delegate to
+     * @param isStartBundle whether or not this is the {@code startBundle} call
```

Comment:
Since it is moderately crufty, might be worth explaining here or somewhere why `isStartBundle` matters. But I, like Luke, don't have a better idea.